### PR TITLE
Feature/cinfra 427 create collection group part3

### PR DESCRIPTION
**Pull request by mchacki**
### Scope & Purpose

*Next step to refactor the create collection. This now does a full road from create-API to agency communication.*
Not all callers are adapted now, this is going to be added in one follow up PR, to avoid increasing this PR even further.

Detailed Workflow description: https://arangodb.atlassian.net/wiki/spaces/DEV/pages/1959395341/Create+Collection+Workflow+replication+2.0

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [x] :hammer: Refactoring/simplification

### Checklist

- [x] Tests
  - [ ] **Regression tests**
  - [x] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [x] Enterprise PR: https://github.com/arangodb/enterprise/pull/1159
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/VocBase/Methods/Collections.cpp [hunk 1]
```diff
@@ -601,37 +603,194 @@ void Collections::enumerate(
...
+    TRI_ASSERT(ServerState::instance()->isSingleServer() ||
+               ServerState::instance()->isDBServer() ||
+               ServerState::instance()->isAgent());
+    // Here we do have a single server setup, or we're either on a DBServer
+    // / Agency. In that case, we're not batching collection creating.
+    // Therefore, we need to iterate over the infoSlice and create each
+    // collection one by one.
+    results = vocbase.createCollections(
+        collections, allowEnterpriseCollectionsOnSingleServer);
+    if (results.fail()) {
```
Comment (mchacki) [inline arangod/VocBase/Methods/Collections.cpp:714]: Add audit logging

File: arangod/Cluster/ClusterCollectionMethods.cpp [hunk 2]
```diff
@@ -0,0 +1,774 @@
...
+Result impl(ClusterInfo& ci, ArangodServer& server,
+            std::string_view databaseName, WriterType& writer,
+            bool waitForSyncReplication) {
+  AgencyComm ac(server);
+  double pollInterval = ci.getPollInterval();
+  AgencyCallbackRegistry& callbackRegistry = ci.agencyCallbackRegistry();
+
+  // TODO Timeout?
+  std::vector<std::string> collectionNames = writer.collectionNames();
+  while (true) {
```
Comment (mchacki) [inline arangod/Cluster/ClusterCollectionMethods.cpp:63]: Maybe we need to have an upper bound on retries

File: arangod/Cluster/Utils/TargetCollectionAgencyWriter.cpp [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+  preconditions =
+      std::move(preconditions)
+          .isIntersectionEmpty(paths::target()->toBeCleanedServers()->str(),
+                               serversPlanned);
+  // Servers should not be in cleaned state
+  preconditions =
+      std::move(preconditions)
+          .isIntersectionEmpty(paths::target()->cleanedServers()->str(),
+                               serversPlanned);
+
```
Comment (mchacki) [inline arangod/Cluster/Utils/TargetCollectionAgencyWriter.cpp:239]: Precondition on health?

**Review by apetenchea — APPROVED** [review]
There will be a few changes required once staging is merged back, mostly because some paths in the agency have been changed in the refactoring branch. This is expected and is not a problem in the actual PR, so I think it makes sense to approve the PR now.
I focused my attention on the `Cluster` and `VocBase` directories, and I went a bit faster over the tests. LGTM!

File: tests/Cluster/PlanCollectionEntryTest.cpp [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  auto distProto = std::make_shared<TestShardDistribution>(
+      numberOfShards, col.replicationFactor.value());
+  auto shards = generateShardNames(numberOfShards);
+  ShardDistribution dist{shards, distProto};
+  AgencyIsBuildingFlags buildingFlags;
+  buildingFlags.rebootId = RebootId(42);
+  buildingFlags.coordinatorName = "CRDN_123";
+  PlanCollectionEntry entry{col, std::move(dist), std::move(buildingFlags)};
+  auto builder = entry.toVPackDeprecated();
+  LOG_DEVEL << builder.toJson();
```
[the remark this answers is not in the document]
Comment (mchacki) [reply tests/Cluster/PlanCollectionEntryTest.cpp:118]: For now yes.
The test has to be added in a follow-up PR.
I dislike right now that we would have to assert on JSON values.
There are tickets for this test in CINFRA board.

File: tests/Cluster/PlanCollectionToAgencyWriterTest.cpp [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+  auto col = prepareMinimalCollection(1, 1);
+
+  auto writer = createWriterWithTestSharding(col);
+
+  auto serversAvailable = generateServerNames(3);
+  auto res =
+      writer.prepareStartBuildingTransaction(dbName(), 2, serversAvailable);
+  ASSERT_TRUE(res.ok());
+  auto transaction = res.get();
+  LOG_DEVEL << transaction.toJson();
```
[the remark this answers is not in the document]
Comment (mchacki) [reply tests/Cluster/PlanCollectionToAgencyWriterTest.cpp:139]: For now yes.
The test has to be added in a follow-up PR.
I dislike right now that we would have to assert on JSON values.
There are tickets for this test in CINFRA board.

File: arangod/Cluster/Utils/TargetCollectionAgencyWriter.cpp [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+          if (report->hasReported(id)) {
+            // This replicatedLog has already reported
+            return true;
+          }
+          if (slice.isNone()) {
+            return false;
+          }
+
+          auto supervision = velocypack::deserialize<
+              replication2::replicated_state::agency::Current::Supervision>(
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Cluster/Utils/TargetCollectionAgencyWriter.cpp]: The code for waitForLogReady is included in a way.
After I rereviewed this piece I think the current version in this PR is too complicated, and can potentially use waitForLogReady code.
Will add a followup ticket on CINFRA board.

File: arangod/Cluster/Utils/CurrentWatcher.cpp [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  _callbacks.emplace_back(std::move(path), std::move(identifier),
+                          std::move(callback));
+  _expectedResults++;
+}
+
+void CurrentWatcher::addReport(std::string identifier, Result result) {
+  _results.doUnderLock([identifier = std::move(identifier),
+                        result = std::move(result)](auto& lists) {
+    if (lists.find(identifier) == lists.end()) {
+      lists.emplace(std::move(identifier), std::move(result));
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Cluster/Utils/CurrentWatcher.cpp:51]: Good point, thank you very much for spotting this!
Applied

File: arangod/Cluster/Utils/DistributeShardsLike.cpp [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+            "Server: " + s +
+                " required to fulfill distributeShardsLike, is not available"};
+      }
+      // Register Server als planned
+      serversPlanned.emplace(s);
+    }
+  }
+
+  // Sharding Okay, take it
+  _shardToServerMapping = nextSharding.get();
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Cluster/Utils/DistributeShardsLike.cpp:61]: yes, added.

File: arangod/Cluster/Utils/PlanCollectionEntry.h [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+#include "ShardDistribution.h"
+#include "Cluster/Utils/PlanShardToServerMappping.h"
+#include "Cluster/Utils/IShardDistributionFactory.h"
+
+namespace arangodb {
+namespace velocypack {
+class Builder;
+}
+
+struct PlanCollectionEntry {
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Cluster/Utils/PlanCollectionEntry.h:39]: Absolutely, however for the time being I could not oversee how different One and Two will be, as soon as we activate CollectionGroups.
If they tend to be equal, we will use inheritance here.

File: arangod/Cluster/Utils/PlanCollectionEntryReplication2.cpp [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+PlanShardToServerMapping PlanCollectionEntryReplication2::getShardMapping()
+    const {
+  TRI_ASSERT(_shardDistribution.getDistributionForShards().shards.size() ==
+             _properties.numberOfShards);
+  return _shardDistribution.getDistributionForShards();
+}
+
+[[nodiscard]] replication2::replicated_state::agency::Target
+PlanCollectionEntryReplication2::getReplicatedStateForTarget(
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Cluster/Utils/PlanCollectionEntryReplication2.cpp]: Change applied

File: arangod/Cluster/ClusterMethods.cpp [hunk 11]
```diff
@@ -2953,13 +2949,6 @@ ClusterMethods::persistCollectionsInAgency(
         if (!avoid.empty()) {
           // We need to remove all servers that are in the avoid list
           if (dbServers.size() - avoid.size() < replicationFactor) {
-            LOG_TOPIC("03682", DEBUG, Logger::CLUSTER)
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Cluster/ClusterMethods.cpp:3001]: I think this was first moved to a different place, so I cut it out here, and copied it over.
However the final variant first removes all avoid servers. And then tests the replicationFactor against remaining servers. => The Log is obsolete in the new variant.

Furthermore this method is the old version, and is to be removed as soon as all parts are moved to new variant.
Also avoid servers is a hidden test-only feature, so not really relevant.

File: arangod/Cluster/Utils/PlanCollectionEntry.cpp [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+#include "VocBase/Properties/CreateCollectionBody.h"
+
+#include <velocypack/Builder.h>
+#include <velocypack/Collection.h>
+
+using namespace arangodb;
+PlanCollectionEntry::PlanCollectionEntry(UserInputCollectionProperties col,
+                                         ShardDistribution shardDistribution,
+                                         AgencyIsBuildingFlags isBuildingFlags)
+    : _properties({col}),
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Cluster/Utils/PlanCollectionEntry.cpp:34]: Moved, good point

File: arangod/VocBase/Methods/Collections.cpp [hunk 13]
```diff
@@ -601,37 +603,194 @@ void Collections::enumerate(
...
+      CollectionNameResolver resolver(vocbase);
+      auto myColToDistributeLike =
+          resolver.getCollection(col.distributeShardsLike.value());
+      if (myColToDistributeLike == nullptr) {
+        return Result{
+            TRI_ERROR_CLUSTER_UNKNOWN_DISTRIBUTESHARDSLIKE,
+            "Collection not found: " + col.distributeShardsLike.value() +
+                " in database " + vocbase.name()};
+      }
+      col.numberOfShards = myColToDistributeLike->numberOfShards();
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/VocBase/Methods/Collections.cpp:650]: They are.
This is the old code variant.
So this is leftover from before, in the old code this is actually the first place doing these checks.

File: arangod/VocBase/Properties/UserInputCollectionProperties.cpp [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+#endif
+  }
+  shardKeys = std::vector<std::string>{StaticStrings::KeyString};
+}
+
+Result UserInputCollectionProperties::validateShardKeys() {
+  // Has to be set by now
+  TRI_ASSERT(shardKeys.has_value());
+
+  // TODO: Copy Pasted Code to validate ShardKeys, maybe move!
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/VocBase/Properties/UserInputCollectionProperties.cpp:247]: This is indeed duplicated code. Removed it.

File: arangod/Cluster/ClusterCollectionMethods.cpp [hunk 15]
```diff
@@ -0,0 +1,774 @@
...
+  // We retain a reference here ourselfs in case we need to retry due to server
+  // failure, this way we can just to create the shards on other servers.
+  std::unordered_map<std::string, std::shared_ptr<IShardDistributionFactory>>
+      shardDistributionList;
+
+  std::vector<EntryType> collectionPlanEntries{};
+  collectionPlanEntries.reserve(collections.size());
+
+  /*
+ NEED TO ACTIVATE THE FOLLOWING CODE
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Cluster/ClusterCollectionMethods.cpp:597]: Still relevant, will address in followup PR.

File: arangod/VocBase/Methods/Collections.cpp (hunk 1 above)
Comment (mchacki) [reply arangod/VocBase/Methods/Collections.cpp:714]: added

**Pull request merged**